lib/common/socket.c

```diff
@@ -1470,3 +1470,112 @@ void h2o_sliding_counter_stop(h2o_sliding_counter_t *counter, uint64_t now)
     /* recalc average */
     counter->average = counter->prev.sum / (sizeof(counter->prev.slots) / sizeof(counter->prev.slots[0]));
 }
+
+#if H2O_USE_DTRACE && defined(__linux__)
+#include <linux/bpf.h>
+#include <linux/unistd.h>
+#include "h2o-probes.h"
+#include "include/h2o/ebpf.h"
+#include <sys/stat.h>
+
+int h2o_tracing_map_fd = 0;
+uint32_t h2o_tracing_map_last_attempt = 0;
+
+void open_map()
+{
+    // only check every second
+    uint32_t now = (uint32_t)time(NULL);
+    if (h2o_tracing_map_last_attempt == now)
+        return;
+
+    __sync_bool_compare_and_swap(&h2o_tracing_map_last_attempt, h2o_tracing_map_last_attempt, now);
+
+    // check if map exists at path
+    struct stat s;
+    if (stat(&H2O_EBPF_MAP_PATH[0], &s) == -1) {
+        // map path unavailable, cleanup fd if needed and leave
+        if (h2o_tracing_map_fd > 0) {
+            close(h2o_tracing_map_fd);
+            __sync_fetch_and_and(&h2o_tracing_map_fd, 0);
+        }
+        return;
+    }
+
+    if (h2o_tracing_map_fd > 0)
+        return; // map still exists and we have a fd
+
+    union bpf_attr attr;
+    memset(&attr, 0, sizeof(attr));
+    attr.pathname = (uint64_t)(unsigned long)&H2O_EBPF_MAP_PATH[0];
+
+    int fd = syscall(__NR_bpf, BPF_OBJ_GET, &attr, sizeof(attr));
+    if (fd <= 0)
+        return;
+
+    // try set map fd if not set before
+    if (!__sync_bool_compare_and_swap(&h2o_tracing_map_fd, 0, fd))
+        close(fd); // if a fd was set beforehand, don't set the new fd, and close it
+}
+
+inline int lookup_map(const void *key, const void *value)
+{
+    union bpf_attr attr;
+    memset(&attr, 0, sizeof(attr));
+    attr.map_fd = h2o_tracing_map_fd;
+    attr.key = (uint64_t)(unsigned long)key;
+    attr.value = (uint64_t)(unsigned long)value;
+    return syscall(__NR_bpf, BPF_MAP_LOOKUP_ELEM, &attr, sizeof(attr)) == -1 ? -1 : 1;
+}
+
+inline void read_ip_port(struct sockaddr *sa, void *ip, uint16_t *port)
```

Comment:
IIUC, this function is essentially buing `h2o_ebpf_map_key_t`. Assuming that is correct, I'd appreciate it if you could change the function signature to something like `init_h2o_ebpf_map_key(h2o_ebpf_map_key_t *key, struct sockaddr *sa)`.